File: ValidBinarySearchTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
//--------------------------
//Space complexity:O(H) and Time complexity: O(n)
class ValidBinarySearchTree {
    public boolean isValidBST(TreeNode root) {
         return inorder(root,null,null);
         
    }
    boolean  inorder(TreeNode root,Integer min,Integer max){
       if(root==null) return true;
        if((min!=null && root.val<=min)|| (max!=null && root.val>=max)){
            return false;
        }
        if(inorder(root.left,min,root.val)==false){
            return false;
        }
        return inorder(root.right,root.val,max);
    }
}

File: Valid_Binary_Search_Tree.cpp
/*
Time Complexity: O(m*n)
Space Complexity: O(m*n)
Did this code successfully run on Leetcode :yes
Any problem you faced while coding this : no
*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    bool isValidBST(TreeNode* root) {
        long preorderI = LONG_MIN;
        stack<TreeNode*> ns;
        while(root!=NULL){
            ns.push(root);
            root=root->left;
        }
        while(!ns.empty()){
            TreeNode *node = ns.top();
            ns.pop();
            //node
            if(node->val <= preorderI){
                return false;
            }
            
            preorderI = node->val;
            //right
            node = node->right;
            while(node!=NULL){
                ns.push(node);
                node=node->left;
            }
        }    
            
    return true;
    }
};